frontend-runitup/src/utils/mapUtils.js

```diff
@@ -95,18 +150,316 @@ export const addRouteToMap = (map, routeGeometry) => {
+    .setHTML(`<h3>Start: ${locationName}</h3>`)
+    .addTo(map);
+
+  return currentMarker;
+};
+
+// Function to clear the route from the map
+export const clearRoute = (map) => {
+  if (map.getSource("route")) {
+    map.removeLayer("route");
+    map.removeSource("route");
+  }
+};
+
+// Function to remove the current marker and popup from the map
+export const removeCurrentMarker = () => {
+  if (currentMarker) {
+    currentMarker.remove();
+    currentMarker = null;
+  }
+  if (currentPopup) {
+    currentPopup.remove();
+    currentPopup = null;
+  }
+};
+
+// Function to calculate estimated running time based on distance
+export const calculateRunningTime = (distanceMiles) => {
+  const averagePaceMinPerMile = 9; // Assume 9 minutes per mile for an average runner
+  const totalMinutes = distanceMiles * averagePaceMinPerMile;
+  const hours = Math.floor(totalMinutes / 60);
+  const minutes = Math.round(totalMinutes % 60);
+  return hours > 0 ? `${hours}h ${minutes}m` : `${minutes}m`;
+};
+
+// Function to get elevation data for a set of coordinates
+export const getElevationData = async (coordinates) => {
+  const chunkSize = 50; // Mapbox allows up to 50 points per request
+  let elevationGain = 0;
+  let elevationLoss = 0;
+  let prevElevation = null;
+
+  for (let i = 0; i < coordinates.length; i += chunkSize) {
+    const chunk = coordinates.slice(i, i + chunkSize);
+    const query = chunk.map((coord) => coord.join(",")).join(",");
+
+    const response = await fetch(
+      `https://api.mapbox.com/v4/mapbox.mapbox-terrain-v2/tilequery/${query}.json?layers=contour&access_token=${mapboxgl.accessToken}`
+    );
+
+    if (!response.ok) {
+      throw new Error(`Failed to fetch elevation data: ${response.statusText}`);
+    }
+
+    const data = await response.json();
+
+    data.features.forEach((feature, index) => {
+      const elevation = feature.properties.ele;
+      if (prevElevation !== null) {
+        const diff = elevation - prevElevation;
+        if (diff > 0) {
+          elevationGain += diff;
+        } else {
+          elevationLoss += Math.abs(diff);
+        }
+      }
+      prevElevation = elevation;
+    });
+  }
+
+  return { gain: Math.round(elevationGain), loss: Math.round(elevationLoss) };
+};
+
+// Function to calculate the distance of a route
+export const calculateRouteDistance = (route) => {
+  return (route.distance / 1609.34).toFixed(2); // Convert meters to miles and round to 2 decimal places
+};
+
+// Function to generate a route within a desired distance
+export const generateRouteWithinDistance = async (
+  startLat,
+  startLng,
+  desiredDistanceMiles,
+  tolerance = 0.1
+) => {
+  const desiredDistanceKm = desiredDistanceMiles * 1.60934;
+  let minRadius = desiredDistanceKm / (4 * Math.PI);
+  let maxRadius = desiredDistanceKm / Math.PI;
+  let bestRoute = null;
+  let bestDistance = Infinity;
+  const maxAttempts = 10;
+
+  // Binary search to find the best route within the desired distance
```

Comment:
I'd suggest creating a new function for the binary search algorithm. This will make it easier for you to add to the binary search algorithm and increase its complexity in the future

Reply:
To the question about the map location before generating a route. Yes i currently gave it a default location to be a random pair of coordinates that happen to be Trenton.

Reply:
For your suggestion would you recommend that i make a file for the whole function, "generateRouteWithinDistance" or more specific for just the binary search?